src/main/scala/org/kframework/kore/interfaces/Interface.scala

```diff
@@ -0,0 +1,303 @@
+package org.kframework.kore.interfaces
+
+object Kore {
+
+  trait Definition {
+    val modules: Seq[Module]
+    val att: Attributes
+  }
+
+  object Definition {
+    def unapply(arg: Definition): Option[(Seq[Module], Attributes)] = Some(arg.modules, arg.att)
+  }
+
+  trait Module {
+    val name: ModuleName
+    val sentences: Seq[Sentence]
+    val att: Attributes
+  }
+
+  object Module {
+    def unapply(arg: Module): Option[(ModuleName, Seq[Sentence], Attributes)] = Some(arg.name, arg.sentences, arg.att)
+  }
+
+  trait Sentence {
+    val att: Attributes
+  }
+
+  trait Import extends Sentence {
+    val name: ModuleName
+  }
+
+  object Import {
+    def unapply(arg: Import): Option[(ModuleName, Attributes)] = Some(arg.name, arg.att)
+  }
+
+  trait SortDeclaration extends Sentence {
+    val sort: Sort
+  }
+
+  object SortDeclaration {
+    def unapply(arg: SortDeclaration): Option[(Sort, Attributes)] = Some(arg.sort, arg.att)
+  }
+
+  trait SymbolDeclaration extends Sentence {
+    val sort: Sort
+    val symbol: Symbol
+    val args: Seq[Sort]
+  }
+
+  object SymbolDeclaration {
+    def unapply(arg: SymbolDeclaration): Option[(Sort, Symbol, Seq[Sort], Attributes)] = Some(arg.sort, arg.symbol, arg.args, arg.att)
+  }
+
+
+  trait Rule extends Sentence {
+    val p: Pattern
+  }
+
+  object Rule {
+    def unapply(arg: Rule): Option[(Pattern, Attributes)] = Some(arg.p, arg.att)
+  }
+
+  trait Axiom extends Sentence {
+    val p: Pattern
+  }
+
+  object Axiom {
+    def unapply(arg: Axiom): Option[(Pattern, Attributes)] = Some(arg.p, arg.att)
+  }
+
+  trait Attributes {
+    def att: Seq[Pattern]
+  }
+
+  object Attributes {
+    def unapply(arg: Attributes): Option[Seq[Pattern]] = Some(arg.att)
+  }
+
+
+  sealed trait Pattern
+
+  trait Variable extends Pattern {
+    val name: Name
+    val sort: Sort
+  }
+
+  object Variable {
+    def unapply(arg: Variable): Option[(Name, Sort)] = Some(arg.name, arg.sort)
+  }
+
+  trait Application extends Pattern {
+    val symbol: Symbol
+    val args: Seq[Pattern]
+  }
+
+  object Application {
+    def unapply(arg: Application): Option[(Symbol, Seq[Pattern])] = Some(arg.symbol, arg.args)
+  }
+
+  trait DomainValue extends Pattern {
+    val symbol: Symbol
+    val value: Value
```

Comment:
It would make my life much easier if `value` would be of type `Any` here. Otherwise, I'd have to wrap the actual value every time, wasting both memory and characters. Safety-wise, we're not gaining much from having this type. For deconstruction, `str` can be replaced with `toString` (which is available in `Any`). For construction, we can have: `DomainValue(Symbol("Int"), "5")`. I don't see `DomainValue(Symbol("Int"), Value("5"))` as providing much/any safety. It gives us nothing at the type level and the `DomainValue` builder itself can do any safety checks on the String.